src/hotspot/share/code/codeBlob.cpp

```diff
@@ -62,32 +62,42 @@ unsigned int CodeBlob::align_code_offset(int offset) {
 }
 
 // This must be consistent with the CodeBlob constructor's layout actions.
-unsigned int CodeBlob::allocation_size(CodeBuffer* cb, int header_size) {
+unsigned int CodeBlob::allocation_size(CodeBuffer* cb, int header_size, bool external_mutable_data) {
   unsigned int size = header_size;
-  size += align_up(cb->total_relocation_size(), oopSize);
+  if (!external_mutable_data) {
+    // In a standard CodeBlob, the relocation_info is located before the code section.
+    // However, in an nmethod, the relocation_info is stored in the mutable data, so
+    // it does not contribute to the nmethod's total size, as it is moved to the C heap.
+    size += align_up(cb->total_relocation_size(), oopSize);
+  }
   // align the size to CodeEntryAlignment
   size = align_code_offset(size);
   size += align_up(cb->total_content_size(), oopSize);
-  size += align_up(cb->total_oop_size(), oopSize);
-  size += align_up(cb->total_metadata_size(), oopSize);
+  if (!external_mutable_data) {
+    size += align_up(cb->total_oop_size(), oopSize);
+    size += align_up(cb->total_metadata_size(), oopSize);
+  }
   return size;
 }
 
 CodeBlob::CodeBlob(const char* name, CodeBlobKind kind, CodeBuffer* cb, int size, uint16_t header_size,
-                   int16_t frame_complete_offset, int frame_size, OopMapSet* oop_maps, bool caller_must_gc_arguments) :
+                   int16_t frame_complete_offset, int frame_size, OopMapSet* oop_maps, bool caller_must_gc_arguments,
+                   bool external_mutable_data) :
```

Comment:
I suggest to add `assert(!external_mutable_data || (kind == CodeBlobKind::Nmethod)`